File: opendc-simulator/opendc-simulator-flow/src/main/java/org/opendc/simulator/flow2/mux/ForwardingFlowMultiplexer.java
/*
 * Copyright (c) 2022 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.simulator.flow2.mux;

import java.util.Arrays;
import java.util.BitSet;
import org.opendc.simulator.flow2.FlowGraph;
import org.opendc.simulator.flow2.FlowStage;
import org.opendc.simulator.flow2.FlowStageLogic;
import org.opendc.simulator.flow2.InHandler;
import org.opendc.simulator.flow2.InPort;
import org.opendc.simulator.flow2.Inlet;
import org.opendc.simulator.flow2.OutHandler;
import org.opendc.simulator.flow2.OutPort;
import org.opendc.simulator.flow2.Outlet;

/**
 * A {@link FlowMultiplexer} implementation that allocates inputs to the outputs of the multiplexer exclusively.
 * This means that a single input is directly connected to an output and that the multiplexer can only support as many
 * inputs as outputs.
 */
public final class ForwardingFlowMultiplexer implements FlowMultiplexer, FlowStageLogic {
    /**
     * Factory implementation for this implementation.
     */
    static FlowMultiplexerFactory FACTORY = ForwardingFlowMultiplexer::new;

    public final IdleInHandler IDLE_IN_HANDLER = new IdleInHandler();
    public final IdleOutHandler IDLE_OUT_HANDLER = new IdleOutHandler();

    private final FlowStage stage;

    private InPort[] inlets;
    private OutPort[] outlets;
    private final BitSet activeInputs;
    private final BitSet activeOutputs;
    private final BitSet availableOutputs;

    private float capacity = 0.f;
    private float demand = 0.f;

    public ForwardingFlowMultiplexer(FlowGraph graph) {
        this.stage = graph.newStage(this);

        this.inlets = new InPort[4];
        this.activeInputs = new BitSet();
        this.outlets = new OutPort[4];
        this.activeOutputs = new BitSet();
        this.availableOutputs = new BitSet();
    }

    @Override
    public float getCapacity() {
        return capacity;
    }

    @Override
    public float getDemand() {
        return demand;
    }

    @Override
    public float getRate() {
        final BitSet activeOutputs = this.activeOutputs;
        final OutPort[] outlets = this.outlets;
        float rate = 0.f;
        for (int i = activeOutputs.nextSetBit(0); i != -1; i = activeOutputs.nextSetBit(i + 1)) {
            rate += outlets[i].getRate();
        }
        return rate;
    }

    @Override
    public int getMaxInputs() {
        return getOutputCount();
    }

    @Override
    public int getMaxOutputs() {
        return Integer.MAX_VALUE;
    }

    @Override
    public int getInputCount() {
        return activeInputs.length();
    }

    @Override
    public Inlet newInput() {
        final BitSet activeInputs = this.activeInputs;
        int slot = activeInputs.nextClearBit(0);

        InPort inPort = stage.getInlet("in" + slot);
        inPort.setMask(true);

        InPort[] inlets = this.inlets;
        if (slot >= inlets.length) {
            int newLength = inlets.length + (inlets.length >> 1);
            inlets = Arrays.copyOf(inlets, newLength);
            this.inlets = inlets;
        }

        final BitSet availableOutputs = this.availableOutputs;
        int outSlot = availableOutputs.nextSetBit(0);

        if (outSlot < 0) {
            throw new IllegalStateException("No capacity available for a new input");
        }

        inlets[slot] = inPort;
        activeInputs.set(slot);

        OutPort outPort = outlets[outSlot];
        availableOutputs.clear(outSlot);

        inPort.setHandler(new ForwardingInHandler(outPort));
        outPort.setHandler(new ForwardingOutHandler(inPort));

        inPort.pull(outPort.getCapacity());

        return inPort;
    }

    @Override
    public void releaseInput(Inlet inlet) {
        InPort port = (InPort) inlet;
        int slot = port.getId();

        final BitSet activeInputs = this.activeInputs;

        if (!activeInputs.get(slot)) {
            return;
        }

        port.cancel(null);
        activeInputs.clear(slot);

        ForwardingInHandler inHandler = (ForwardingInHandler) port.getHandler();
        availableOutputs.set(inHandler.output.getId());

        port.setHandler(IDLE_IN_HANDLER);
    }

    @Override
    public int getOutputCount() {
        return activeOutputs.length();
    }

    @Override
    public Outlet newOutput() {
        final BitSet activeOutputs = this.activeOutputs;
        int slot = activeOutputs.nextClearBit(0);

        OutPort port = stage.getOutlet("out" + slot);
        OutPort[] outlets = this.outlets;
        if (slot >= outlets.length) {
            int newLength = outlets.length + (outlets.length >> 1);
            outlets = Arrays.copyOf(outlets, newLength);
            this.outlets = outlets;
        }
        outlets[slot] = port;

        activeOutputs.set(slot);
        availableOutputs.set(slot);

        port.setHandler(IDLE_OUT_HANDLER);

        return port;
    }

    @Override
    public void releaseOutput(Outlet outlet) {
        OutPort port = (OutPort) outlet;
        int slot = port.getId();
        activeInputs.clear(slot);
        availableOutputs.clear(slot);
        port.complete();

        port.setHandler(IDLE_OUT_HANDLER);
    }

    @Override
    public long onUpdate(FlowStage ctx, long now) {
        return Long.MAX_VALUE;
    }

    class ForwardingInHandler implements InHandler {
        final OutPort output;

        ForwardingInHandler(OutPort output) {
            this.output = output;
        }

        @Override
        public float getRate(InPort port) {
            return output.getRate();
        }

        @Override
        public void onPush(InPort port, float rate) {
            ForwardingFlowMultiplexer.this.demand += -port.getDemand() + rate;

            output.push(rate);
        }

        @Override
        public void onUpstreamFinish(InPort port, Throwable cause) {
            ForwardingFlowMultiplexer.this.demand -= port.getDemand();

            final OutPort output = this.output;
            output.push(0.f);

            releaseInput(port);
        }
    }

    private class ForwardingOutHandler implements OutHandler {
        private final InPort input;

        ForwardingOutHandler(InPort input) {
            this.input = input;
        }

        @Override
        public void onPull(OutPort port, float capacity) {
            ForwardingFlowMultiplexer.this.capacity += -port.getCapacity() + capacity;

            input.pull(capacity);
        }

        @Override
        public void onDownstreamFinish(OutPort port, Throwable cause) {
            ForwardingFlowMultiplexer.this.capacity -= port.getCapacity();

            input.cancel(cause);

            releaseOutput(port);
        }
    }

    private static class IdleInHandler implements InHandler {
        @Override
        public float getRate(InPort port) {
            return 0.f;
        }

        @Override
        public void onPush(InPort port, float rate) {
            port.cancel(new IllegalStateException("Inlet is not allocated"));
        }

        @Override
        public void onUpstreamFinish(InPort port, Throwable cause) {}
    }

    private class IdleOutHandler implements OutHandler {
        @Override
        public void onPull(OutPort port, float capacity) {
            ForwardingFlowMultiplexer.this.capacity += -port.getCapacity() + capacity;
        }

        @Override
        public void onDownstreamFinish(OutPort port, Throwable cause) {
            ForwardingFlowMultiplexer.this.capacity -= port.getCapacity();
        }
    }
}

File: opendc-simulator/opendc-simulator-flow/src/main/kotlin/org/opendc/simulator/flow/mux/ForwardingFlowMultiplexer.kt
/*
 * Copyright (c) 2021 AtLarge Research
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package org.opendc.simulator.flow.mux

import org.opendc.simulator.flow.*
import org.opendc.simulator.flow.interference.InterferenceKey
import java.util.ArrayDeque
import kotlin.math.max

/**
 * A [FlowMultiplexer] implementation that allocates inputs to the outputs of the multiplexer exclusively. This means
 * that a single input is directly connected to an output and that the multiplexer can only support as many
 * inputs as outputs.
 *
 * @param engine The [FlowEngine] driving the simulation.
 * @param listener The convergence listener of the multiplexer.
 */
public class ForwardingFlowMultiplexer(
    private val engine: FlowEngine,
    private val listener: FlowConvergenceListener? = null
) : FlowMultiplexer, FlowConvergenceListener {
    override val inputs: Set<FlowConsumer>
        get() = _inputs
    private val _inputs = mutableSetOf<Input>()

    override val outputs: Set<FlowSource>
        get() = _outputs
    private val _outputs = mutableSetOf<Output>()
    private val _availableOutputs = ArrayDeque<Output>()

    override val counters: FlowCounters = object : FlowCounters {
        override val demand: Double
            get() = _outputs.sumOf { it.forwarder.counters.demand }
        override val actual: Double
            get() = _outputs.sumOf { it.forwarder.counters.actual }
        override val remaining: Double
            get() = _outputs.sumOf { it.forwarder.counters.remaining }
        override val interference: Double
            get() = _outputs.sumOf { it.forwarder.counters.interference }

        override fun reset() {
            for (output in _outputs) {
                output.forwarder.counters.reset()
            }
        }

        override fun toString(): String = "FlowCounters[demand=$demand,actual=$actual,remaining=$remaining]"
    }

    override val rate: Double
        get() = _outputs.sumOf { it.forwarder.rate }

    override val demand: Double
        get() = _outputs.sumOf { it.forwarder.demand }

    override val capacity: Double
        get() = _outputs.sumOf { it.forwarder.capacity }

    override fun newInput(key: InterferenceKey?): FlowConsumer {
        val output = checkNotNull(_availableOutputs.poll()) { "No capacity to serve request" }
        val input = Input(output)
        _inputs += input
        return input
    }

    override fun newInput(capacity: Double, key: InterferenceKey?): FlowConsumer = newInput(key)

    override fun removeInput(input: FlowConsumer) {
        if (!_inputs.remove(input)) {
            return
        }

        val output = (input as Input).output
        output.forwarder.cancel()
        _availableOutputs += output
    }

    override fun newOutput(): FlowSource {
        val forwarder = FlowForwarder(engine, this)
        val output = Output(forwarder)

        _outputs += output
        return output
    }

    override fun removeOutput(output: FlowSource) {
        if (!_outputs.remove(output)) {
            return
        }

        val forwarder = (output as Output).forwarder
        forwarder.close()
    }

    override fun clearInputs() {
        for (input in _inputs) {
            val output = input.output
            output.forwarder.cancel()
            _availableOutputs += output
        }

        _inputs.clear()
    }

    override fun clearOutputs() {
        for (output in _outputs) {
            output.forwarder.cancel()
        }
        _outputs.clear()
        _availableOutputs.clear()
    }

    override fun clear() {
        clearOutputs()
        clearInputs()
    }

    private var _lastConverge = Long.MAX_VALUE

    override fun onConverge(now: Long, delta: Long) {
        val listener = listener
        if (listener != null) {
            val lastConverge = _lastConverge
            _lastConverge = now
            val duration = max(0, now - lastConverge)
            listener.onConverge(now, duration)
        }
    }

    /**
     * An input on the multiplexer.
     */
    private inner class Input(@JvmField val output: Output) : FlowConsumer by output.forwarder {
        override fun toString(): String = "ForwardingFlowMultiplexer.Input"
    }

    /**
     * An output on the multiplexer.
     */
    private inner class Output(@JvmField val forwarder: FlowForwarder) : FlowSource by forwarder {
        override fun onStart(conn: FlowConnection, now: Long) {
            _availableOutputs += this
            forwarder.onStart(conn, now)
        }

        override fun onStop(conn: FlowConnection, now: Long, delta: Long) {
            forwarder.cancel()
            forwarder.onStop(conn, now, delta)
        }

        override fun toString(): String = "ForwardingFlowMultiplexer.Output"
    }
}
